### PR TITLE
refactor firebase functions

### PR DESCRIPTION
**Pull request by imeha**

### CODE REVIEW

File: firebase-functions/functions/src/operations/GameOperations.ts [hunk 1]
```diff
@@ -8,30 +8,38 @@ import { Result } from '../models/result';
...
 import { Game } from '../models/game';
 import { collections } from '../constants/collections';
-import { Room } from '../models/room';
-import { GameType } from '../models/game-type';
 import DocumentData = firestore.DocumentData;
 import QueryDocumentSnapshot = firestore.QueryDocumentSnapshot;
+import { getRoom } from './roomOperations';
 
-export async function addRoundToGame(roomId: string, players: QueryDocumentSnapshot<DocumentData>[]): Promise<void> {
+export async function getCurrentGame(roomId: string, players: QueryDocumentSnapshot<DocumentData>[]): Promise<Game> {
```
Comment (DanielMuckelbauer) [inline firebase-functions/functions/src/operations/GameOperations.ts:15]: I think this method name is really misleading. It doesn't merely get the game, it adds a round to the current game or creates a new one. 
I think the old name was better. I suggest you either:
- create a second method that returns the current game: this keeps the method signature honest and splits between methods that change state and methods that return something (CQRS)
- rename it back and just still return the game: I think that is the worse solution, but less convoluted

File: firebase-functions/functions/src/firebase/calculateResult.ts [hunk 2]
```diff
@@ -1,13 +1,10 @@
...
-import { Player } from '../models/player';
 import { hasEveryoneChosen } from '../operations/choiceOperations';
-import { addRoundToGame } from '../operations/GameOperations';
-import { collections } from '../constants/collections';
-import { deactivatePlayers, resetPlayerChoices } from '../operations/playerOperations';
-import QueryDocumentSnapshot = firestore.QueryDocumentSnapshot;
-import DocumentData = firestore.DocumentData;
+import { deactivatePlayers, getActivePlayersInRoom, resetPlayerChoices } from '../operations/playerOperations';
+import { getCurrentRoomId } from '../operations/roomOperations';
+import { getCurrentGame } from '../operations/GameOperations';
```
Comment (DanielMuckelbauer) [inline firebase-functions/functions/src/firebase/calculateResult.ts:6]: Could you rename this file to be lowercase, like the others? :)

File: firebase-functions/functions/src/operations/GameOperations.ts [hunk 3]
```diff
@@ -8,30 +8,38 @@ import { Result } from '../models/result';
...
   } else {
-    games.push(createNewGame(round, room.typeOfGame));
+    games[games.length - 1].rounds.push(round);
   }
   await admin.firestore().collection(collections.rooms).doc(roomId).update({ games });
+  return games[games.length - 1];
 }
 
-function getPlayerChoices(players: QueryDocumentSnapshot<DocumentData>[]): PlayerChoice[] {
+export async function addLastOneActiveToGame(lastOneActive: string, roomId: string): Promise<void> {
```
Comment (DanielMuckelbauer) [inline firebase-functions/functions/src/operations/GameOperations.ts:34]: Not to be picky, but `set...` might be preferable? After all, you set a property. Add implies there is some list involved.

File: firebase-functions/functions/src/operations/playerOperations.ts [hunk 4]
```diff
@@ -1,32 +1,57 @@
...
-async function determineLastOneActive(initiallyActivePlayers: QueryDocumentSnapshot<DocumentData>[], playersToDeactivate: QueryDocumentSnapshot<DocumentData>[], roomId: string): Promise<void> {
-  const stillActivePlayers = initiallyActivePlayers.filter(player => !playersToDeactivate.includes(player));
-  if (stillActivePlayers.length === 1) {
-    await admin.firestore().collection(collections.rooms).doc(roomId).update({ lastOneActive: stillActivePlayers[0].id });
-    await updatePlayers(stillActivePlayers, { isObserver: true });
+async function determineLastOneActive(initiallyActivePlayers: QueryDocumentSnapshot<DocumentData>[], playersToDeactivate: string[], roomId: string): Promise<void> {
+  if (initiallyActivePlayers.length === playersToDeactivate.length + 1) {
+    const lastOneActive = initiallyActivePlayers.find(player => !playersToDeactivate.includes(player.id));
+    if (lastOneActive) {
+      await addLastOneActiveToGame(lastOneActive.id, roomId);
```
Comment (DanielMuckelbauer) [inline firebase-functions/functions/src/operations/playerOperations.ts:35]: Is this check necessary, or does it maybe even hide a bug? 
If we get to that line and we didn't find a lastOneActive, something is wrong I think.

File: firebase-functions/functions/src/operations/playerOperations.ts [hunk 5]
```diff
@@ -1,32 +1,57 @@
...
+  await updatePlayersDoc(players, { choice: null });
 }
 
-async function determineLastOneActive(initiallyActivePlayers: QueryDocumentSnapshot<DocumentData>[], playersToDeactivate: QueryDocumentSnapshot<DocumentData>[], roomId: string): Promise<void> {
-  const stillActivePlayers = initiallyActivePlayers.filter(player => !playersToDeactivate.includes(player));
-  if (stillActivePlayers.length === 1) {
-    await admin.firestore().collection(collections.rooms).doc(roomId).update({ lastOneActive: stillActivePlayers[0].id });
-    await updatePlayers(stillActivePlayers, { isObserver: true });
+async function determineLastOneActive(initiallyActivePlayers: QueryDocumentSnapshot<DocumentData>[], playersToDeactivate: string[], roomId: string): Promise<void> {
+  if (initiallyActivePlayers.length === playersToDeactivate.length + 1) {
```
Comment (DanielMuckelbauer) [inline firebase-functions/functions/src/operations/playerOperations.ts:32]: Smart, but I suggest extracting a variable `isOnlyOnePlayerLeft` or so

File: firebase-functions/functions/src/operations/playerOperations.ts [hunk 6]
```diff
@@ -1,32 +1,57 @@
...
+import { Game } from '../models/game';
+import { GameType } from '../models/game-type';
+import { Result } from '../models/result';
+import { getCurrentGameType } from './roomOperations';
+import { addLastOneActiveToGame } from './GameOperations';
 
-export async function deactivatePlayers(roomId: string, initiallyActivePlayers: QueryDocumentSnapshot<DocumentData>[]): Promise<void> {
-  const playersToDeactivate = await getPlayersToDeactivate(initiallyActivePlayers, roomId);
+export async function deactivatePlayers(roomId: string, game: Game, initiallyActivePlayers: QueryDocumentSnapshot<DocumentData>[]): Promise<void> {
+  const playersToDeactivate = await getPlayersToDeactivate(game, roomId);
```
Comment (DanielMuckelbauer) [inline firebase-functions/functions/src/operations/playerOperations.ts:14]: Maybe rename this to `playerIdsToDeactivate` (same in method). Little more precise?

File: firebase-functions/functions/src/operations/roomOperations.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import { GameType } from '../models/game-type';
+import * as admin from 'firebase-admin';
+import { collections } from '../constants/collections';
+import { Room } from '../models/room';
+
+export async function getCurrentGameType(roomId: string): Promise<GameType | undefined> {
+  const room = await admin.firestore().collection(collections.rooms).doc(roomId).get();
+  return (room.data() as Room).typeOfGame;
```
Comment (DanielMuckelbauer) [inline firebase-functions/functions/src/operations/roomOperations.ts:8]: You could call below method and avoid duplication:

`  return (await getRoom(roomId)).typeOfGame;`

File: firebase-functions/functions/src/models/round.ts [hunk 8]
```diff
@@ -1,6 +1,6 @@
 ﻿import { PlayerChoice } from './player-choice';
 
 export interface Round {
-  playerChoices: PlayerChoice[];
+  playersChoices: PlayerChoice[];
   timeStamp: number;
```
Comment (DanielMuckelbauer) [inline firebase-functions/functions/src/models/round.ts:5]: Much better in fact :)

**Review by DanielMuckelbauer — CHANGES_REQUESTED** [review]

File: firebase-functions/functions/src/operations/playerOperations.ts (hunk 4 above)
Comment (imeha) [reply firebase-functions/functions/src/operations/playerOperations.ts:35]: Logically, it won't happen but the method `find `by definition return undefined that should be guarded. Anyway, I removed the guard to another level.

File: firebase-functions/functions/src/operations/GameOperations.ts (hunk 1 above)
Comment (imeha) [reply firebase-functions/functions/src/operations/GameOperations.ts:15]: About your:
1st suggestion: one of the main focuses in this refactoring is to reduce as possible firebase DB calls, spliting means different calls as we don't have a shared store.
2nd suggestion: `addRoundToGame`  and `getCurrentGame` as equally convoluted :), nothing makes it less convoluted :)

I have changed it tho to `addRoundToGame` so I won't hold this PR for long, until we find a more suitable name.

Comment (DanielMuckelbauer) [reply firebase-functions/functions/src/operations/GameOperations.ts]: Thanks, you are right with that first argument.

About the second: the problem with `getCurrentGame` is that it doesn't indicate that something _changes_. You call it and expect to get a game, but in reality you add a round to a game and even create a new one possibly. That is some thick side effect for a `get...` method. 

Anyway, that was just for completeness :) Changing the name back works for me, thanks.

**Review by DanielMuckelbauer — APPROVED** [review]
Good stuff

**Pull request merged**